jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUForestIsomorphismInspector.java

```diff
@@ -0,0 +1,236 @@
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+
+        if (!forest2.vertexSet().containsAll(roots2)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+    }
+
+    private V getFreshVertex(Graph<V, E> graph){
+        Supplier<V> supplier = graph.getVertexSupplier();
+
+        if (Objects.isNull(supplier))
+            throw new IllegalArgumentException("vertex supplier cannot be null");
+
+        V v = supplier.get();
+        assert !graph.vertexSet().contains(v);
+        return v;
+    }
+
+    private Graph<V, E> takeSubgraph(Graph<V, E> graph, Set<V> vertices){
+        Graph<V, E> subgraph = new SimpleGraph<>(graph.getVertexSupplier(), graph.getEdgeSupplier(), false);
+
+        for (V v: vertices)
+            subgraph.addVertex(v);
+
+        for (V v: vertices){
+            for (E edge: graph.edgesOf(v)){
+                V u = Graphs.getOppositeVertex(graph, edge, v);
+
+                if (vertices.contains(u))
+                    subgraph.addEdge(u, v);
+            }
+        }
+
+        return subgraph;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<GraphMapping<V, E>> getMappings() {
+        GraphMapping<V, E> iterMapping = getMapping();
+
+        if (iterMapping == null)
+            return Collections.emptyIterator();
+        else
+            return Collections.singletonList(iterMapping).iterator();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isomorphismExists(){
+        if (!computed){
+            isomorphicMapping = getMapping();
+        }
+
+        return isomorphicMapping != null;
+    }
+
+    private Pair<V, Graph<V, E>> addDummyRoot(Graph<V, E> forest, Set<V> roots){
+        Graph<V, E> freshForest =
+                new SimpleGraph<>(forest.getVertexSupplier(), forest.getEdgeSupplier(), false);
+
+        roots.forEach(freshForest::addVertex);
+        V freshVertex = getFreshVertex(forest);
+
+        freshForest.addVertex(freshVertex);
+
+        for (V root: roots)
+            freshForest.addEdge(freshVertex, root);
+
+        return Pair.of(freshVertex, new AsGraphUnion<>(freshForest, forest));
+    }
+
+    /**
+     * Get an isomorphism between the input forest or {@code null} if none exists.
+     *
+     * @return isomorphic mapping, {@code null} is none exists
+     */
+    public IsomorphicGraphMapping<V, E> getMapping(){
+        if (computed) {
+            return isomorphicMapping;
+        }
+
+        ConnectivityInspector<V, E> connectivityInspector1 = new ConnectivityInspector<>(forest1);
+        List<Set<V>> trees1 = connectivityInspector1.connectedSets();
+
+        ConnectivityInspector<V, E> connectivityInspector2 = new ConnectivityInspector<>(forest2);
+        List<Set<V>> trees2 = connectivityInspector2.connectedSets();
+
+        if (trees1.size() <= 1 && trees2.size() <= 1) {
```

Comment:
Now that we no longer need trees1 and trees2 for the else branch, do you think detecting this special case is still worth the cost of the connectivity inspectors?  Or if so, can't it be detected just by the number of roots provided?